src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,471 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.crypto.provider;
+
+import javax.crypto.KDFSpi;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.HKDFParameterSpec;
+import javax.crypto.spec.KDFParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.ProviderException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidParameterSpecException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * KeyDerivation implementation for the HKDF function.
+ * <p>
+ * This class implements the HKDF-Extract and HKDF-Expand functions from RFC 5869.  This
+ * implementation provides the complete Extract-then-Expand HKDF function as well as Extract-only
+ * and Expand-only variants.
+ */
+abstract class HkdfKeyDerivation extends KDFSpi {
+
+    protected Mac hmacObj;
+    protected int hmacLen;
+    protected String hmacAlgName;
+    protected List<SecretKey> ikms;
+    protected List<SecretKey> salts;
+    protected SecretKey initialKeyMaterial;
+    protected SecretKey salt;
+    protected SecretKey pseudoRandomKey;
+    protected byte[] info;
+    protected int length;
+
+    protected enum HKDFTYPES {
+        EXTRACT, EXPAND, EXTRACTEXPAND
+    }
+
+    protected HKDFTYPES HKDFTYPE;
+
+    /**
+     * The sole constructor.
+     *
+     * @param algParameterSpec
+     *     the initialization parameters (may be {@code null})
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this {@code KDFSpi}
+     */
+    protected HkdfKeyDerivation(AlgorithmParameterSpec algParameterSpec)
+        throws InvalidAlgorithmParameterException {
+        super(algParameterSpec);
+    }
+
+    /**
+     * TODO: description
+     *
+     * @return a derived {@code SecretKey} object of the specified algorithm
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the current {@code KDFParameterSpec} is invalid or
+     *     incorrect for the type of key to be derived
+     * @throws IllegalStateException
+     *     if the key derivation implementation cannot support additional calls to
+     *     {@code deriveKey}
+     */
+    @Override
+    protected SecretKey engineDeriveKey(String alg, KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+
```

Comment:
Too many duplicate code with `engineDerivedata`. Please consider let one call the other, or, you can create a `deriveInternal` method which is then called by the 2 engine methods.

If you do this, there is also no need for a separate `inspectKDFParameterSpec` method, which then make it easy to only have local variables.